### DISCUSSION TITLE
High volumes of data - best practices

### DISCUSSION THREAD

**Discussion by ziverlichson** [TanStack/table#2392] [General]
Hi everyone,
What are the best practices to follow when dealing with high volumes of data? For example, 1 milion entries (rows).

Pagination is probably one answer, but what else?
How would you handle actions performed on the entire set of data, such as sorting/filtering?

Thanks.

**Comment by sbimochan** [comment]
Enable paginations with dropdown giving option for number of rows. 
Sort/Filter only the result shown by pagination. For example 10/1M. It will sort only among 10 or 100 selected by user.
Add a search too with asyncDebounce of around 200ms to prevent high CPU usage.

**Reply by ziverlichson** [reply]
I have enabled pagination, but in my case, sorting & filtering are tabular operations - meaning, they are performed on **all** rows in the table, not only those on the current page. For example, if I sort by 'name', the last row on the last page of the table, might become the first row on the first page.

**Comment by gargroh** [comment]
One suggestion could be - Try to reduce cost of render...

**Comment by hkurra** [comment]
In one of my use-case, we are using react-virtual to virtualize the rows, [codesandbox here](https://codesandbox.io/s/react-table-virtual-wg3ki).
I have also enabled Lazy-loading to load more data when user scrolled to the end only.
Now coming to sorting, I am using server side sorting which means, If user change the sorting, i will make query for DB and get few initial records(You can say page1) then user can keep scrolling with this filter on

**Reply by ziverlichson** [reply]
Thanks @hkurra .
Were you able to work use react-virtual with dynamic sized rows or rows that can expand?

**Reply by hkurra** [reply]
Hmmm, in our case rows are of fixed size but i am passing ref of the `row`  to the react-virtual item through `measureRef` property, as per my understanding, it is meant for same purpose which is to calculate the dimension of row dynamically based from ref.
More info [here](https://www.npmjs.com/package/react-virtual#returns)
![image](https://user-images.githubusercontent.com/6569316/94125152-a0c23e00-fe73-11ea-8dfd-3ecae4b92358.png)
You can also provide `estimateSize` option to react-virtual, but i dont this will be suitable in your case
![image](https://user-images.githubusercontent.com/6569316/94125285-c9e2ce80-fe73-11ea-8016-6d121dfdcbef.png)

**Reply by danvln** [reply]
Hey folks, 

I am facing the same challenge. I am looking for a solution to use expand with react-window/VariableSizeList or Fixed. Did you manage to get anything working? I would appreciate any pointers. Please share if you have any example. 

Thanks!

**Reply by hkurra** [reply]
@danvln I am using react-virtual for virtualization and also tried react-window for virtualization it also works fine but I was facing some issues with the variable Height that's what I switch back to the react-virtual, I am also using row expansion with the react-virtual without any major issues as of now, although the changes are still in development still, things are good enough
